crates/router/src/core/user.rs

```diff
@@ -1856,15 +1865,26 @@ pub async fn verify_recovery_code(
         return Err(UserErrors::TwoFactorAuthNotSetup.into());
     }
 
+    if tfa_utils::get_recovery_code_attempts_from_redis(&state, &user_token.user_id).await?
+        >= consts::user::RECOVERY_CODE_MAX_ATTEMPTS
+    {
+        return Err(UserErrors::MaxRecoveryCodeAttemptsReached.into());
+    }
+
     let mut recovery_codes = user_from_db
         .get_recovery_codes()
         .ok_or(UserErrors::InternalServerError)?;
 
-    let matching_index = utils::user::password::get_index_for_correct_recovery_code(
+    let Some(matching_index) = utils::user::password::get_index_for_correct_recovery_code(
         &req.recovery_code,
         &recovery_codes,
     )?
-    .ok_or(UserErrors::InvalidRecoveryCode)?;
+    else {
+        let _ = tfa_utils::insert_recovery_code_attempts_in_redis(&state, &user_token.user_id)
```

Comment:
`get_recovery_code_attempts_from_redis` is also being called from `insert_recovery_code_attempts_in_redis`.